src/Shared/PackageManagers/NuGetProjectManager.cs

```diff
@@ -7,16 +7,66 @@
 using System.Linq;
 using System.Text.Json;
 using System.Threading.Tasks;
+using System.Xml.Linq;
 
 namespace Microsoft.CST.OpenSource.Shared
 {
     internal class NuGetProjectManager : BaseProjectManager
     {
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Style", "IDE0044:Add readonly modifier", Justification = "Modified through reflection.")]
         public static string ENV_NUGET_ENDPOINT_API = "https://api.nuget.org";
+        
+        private readonly string NUGET_DEFAULT_REGISTRATION_ENDPOINT = "https://api.nuget.org/v3/registration5-gz-semver2/";
+        
+        private string? RegistrationEndpoint { get; set; } = null;
 
         public NuGetProjectManager(string destinationDirectory) : base(destinationDirectory)
         {
+            GetRegistrationEndpointAsync().Wait();
+        }
+
+        /// <summary>
+        /// Dynamically identifies the registration endpoint.
+        /// </summary>
+        /// <returns>NuGet registration endpoint</returns>
+        private async Task<string> GetRegistrationEndpointAsync()
+        {
+            if (RegistrationEndpoint != null)
+            {
+                return RegistrationEndpoint;
+            }
+            
+            try
+            {
+                var doc = await GetJsonCache($"{ENV_NUGET_ENDPOINT_API}/v3/index.json");
+                var resources = doc.RootElement.GetProperty("resources").EnumerateArray();
+                foreach (var resource in resources)
+                {
+                    try
+                    {
+                        var _type = resource.GetProperty("@type").GetString();
+                        if (_type != null && _type.Equals("RegistrationsBaseUrl/Versioned", StringComparison.InvariantCultureIgnoreCase))
+                        {
+                            var _id = resource.GetProperty("@id").GetString();
+                            if (!string.IsNullOrWhiteSpace(_id))
+                            {
+                                RegistrationEndpoint = _id;
+                                return _id;
+                            }
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        Logger.Warn(ex, "Error parsing NuGet API endpoint: {0}", ex.Message);
+                    }
+                }
+            }
+            catch(Exception ex)
+            {
+                Logger.Warn(ex, "Error parsing NuGet API endpoint: {0}", ex.Message);
+            }
+            RegistrationEndpoint = NUGET_DEFAULT_REGISTRATION_ENDPOINT;
+            return NUGET_DEFAULT_REGISTRATION_ENDPOINT;
```

Comment:
shouldn't it be return RegistrationEndpoint;?

Reply:
Shouldn't matter - they're the same value.

Reply:
I recommend changing it in case the previous line changes.

Reply:
Thanks! Fixed.